doc/specs/stdlib_hash_functions.md

```diff
@@ -0,0 +1,1696 @@
+
+
+#### `UNIVERSAL_MULT_HASH` - maps an integer to a smaller number of bits
+
+##### Status
+
+Experimental
+
+##### Description
+
+Calculates an `nbits` hash code from a 32 bit integer.
+
+##### Syntax
+
+`code = [[stdlib_32_bit_hash_codes:universal_mult_hash]]( key, seed, nbits )`
+
+##### Class
+
+Pure function
+
+##### Arguments
+
+`key`: Shall be a scalar integer expression of kind `INT32`. It is an
+`intent(in)` argument.
+
+`seed`: Shall be a scalar integer expression of kind `INT32`. It is an
+`intent(in)` argument. It must have an odd value.
+
+`nbits` Shall be a scalar default integer expression with `0 < nbits <
+32`. It is an `intent(in)` argument.
+
+##### Result
+
+The result is a scalar integer of kind `INT32` with at most the lowest
+`nbits` nonzero.
+
+##### Note
+
+`UNIVERSAL_MULT_HASH` is an implementation of the Universal
+Multiplicative Hash of M. Dietzfelbinger, et al.
+It multiplies the `KEY` by `SEED`, and returns the
+`NBITS` upper bits of the product as the lowest bits of the result.
+
+##### Example
+
+```fortran
+    program demo_universal_mult_hash
+      use stdlib_32_bit_hash_codes, only: odd_random_integer, &
+	  universal_mult_hash
+      use iso_fortran_env, only: int32 
+      implicit none
+      integer, allocatable :: array1(:)
+      integer(int32) :: hash, i, seed, source
+      seed = 0
+      allocate( array1(0:2**6-1) )
+      do i = 0, 2**6-1
+          array(i) = i
+      end do
+      call odd_random_integer( seed )
+      source = int(Z'1FFFFFF', int32)
+      hash = universal_mult_hash(source, seed, 6)
+      azray1(hash) = source
+      print *, seed, hash, array1
+    end program demo_odd_random_integer
+```
+
+#### `WATER_HASH`- calculates a hash code from a key and a seed
+
+##### Status
+
+Experimental
+
+##### Description
+
+Calculates a 32 bit hash code from a rank 1 integer array or a default
+character string, and the input `seed`.
+
+##### Syntax
+
+`code = [[stdlib_32_bit_hash_codes:water_hash]]( key, seed )`
+
+##### Class
+
+Pure function
+
+##### Arguments
+
+`key`: Shall be a deferred length default character scalar expression
+or a rank 1 integer array expression of kind `INT8`, `INT16`,
+`INT32`, or `INT64`.
+It is an `intent(in)` argument.
+
+`seed`: shall be an integer scalar expression of kind `INT64`.
+It is an `intent(in)` argument.
+
+##### Result
+
+The result is a scalar integer of kind `INT32`.
+
+##### Note
+
+`WATER_HASH` is an implementation of the `waterhash` hash code of
+Tommy Ettinger.
+This code has excellent performance on long keys, and good performance
+on short keys.
+As a result it should give reasonable performance for typical hash
+table applications.
+This code passes the SMHasher tests.
+The `waterhash` is based on the `wyhash` of Wang Yi.
+While `wyhash` has a number of bad seeds, depending on the version,
+so far testing has not found any bad seeds for `waterhash`.
+It can have undefined behavior if the key is not word aligned. 
```

Comment:
Changed to:

The `waterhash` is based on the `wyhash` of Wang Yi.
While `wyhash` has a number of bad seeds, where randomiaation of the
output is poor,
so far testing has not found any bad seeds for `waterhash`.
It can have undefined behavior if the key is not word aligned,
i.e. some computer processors can only process a given size integer if
the address of the integer is a multiple of the integer size.